File: apps/web/src/model/layers.ts
import { Emitter } from '../core/emitter';
import { Signal } from '../core/signal';
import type { LayerRenderer } from './style';

export type LineType = 'continuous' | 'dashed' | 'dashdot' | 'dotted';

export type PointSymbol = 'ring' | 'cross' | 'triangle';

/** How entity labels on a layer are drawn. Screen sizes are CSS px. */
export interface LabelStyle {
  /** centre of a polygon, top-left of its bounds, beside a point, or along a line. */
  placement: 'center' | 'corner' | 'beside' | 'along';
  size: number;
  /** Extra px per (px/m) of zoom, capped at maxSize. */
  grow?: number;
  maxSize?: number;
  weight?: 400 | 500 | 600;
  /** "{label}" is replaced with the entity label. */
  template?: string;
  /** Hide when the feature's smaller side is below this on screen. */
  minFeaturePx?: number;
  /** Visible zoom range in px per metre. */
  minScale?: number;
  maxScale?: number;
  ink?: 'fg' | 'fg-dim' | 'label';
}

export interface LayerStyle {
  /** Hex colour, or a theme token: "fg" (main ink) / "fg-dim". */
  color: string;
  lineType: LineType;
  /** Plot line weight in mm. */
  lineWeight: number;
  /** Optional polygon fill (hex with alpha, e.g. "#7FB2E526"). */
  fill?: string;
  /** Symbol and diameter (px) for point entities. */
  point?: { symbol: PointSymbol; size: number };
  label?: LabelStyle;
  /**
   * Whether clicking inside a closed shape selects it (GIS-style area pick).
   * Off for frames and reference outlines that enclose everything.
   */
  pickInterior?: boolean;
  /**
   * The layer's style (docs/STYLE.md): which symbols its objects get. When
   * absent, the simple look above (colour, line type, weight, fill) is used.
   */
  renderer?: LayerRenderer;
}

export interface LayerNode {
  id: string;
  name: string;
  type: 'group' | 'layer';
  visible: boolean;
  locked: boolean;
  expanded: boolean;
  style: LayerStyle;
  children: LayerNode[];
}

export type LayerInit = Partial<Omit<LayerNode, 'children' | 'style'>> & {
  name: string;
  style?: Partial<LayerStyle>;
  children?: LayerInit[];
};

export const LINE_TYPE_LABEL: Record<LineType, string> = {
  continuous: 'Sürekli',
  dashed: 'Kesikli',
  dashdot: 'Noktalı kesik',
  dotted: 'Noktalı',
};

interface LayerEvents {
  /** Structure changed (add/remove/rename/reorder). */
  structure: void;
  /** Visibility / lock / style changed — carries affected leaf layer ids. */
  state: { ids: string[] };
  /** A group was opened or closed in the tree: a view change, not an edit. */
  expanded: { id: string };
}

let uid = 0;
const defaultStyle: LayerStyle = { color: 'fg', lineType: 'continuous', lineWeight: 0.18 };

export class LayerStore {
  readonly events = new Emitter<LayerEvents>();
  /** Bumped on any change — cheap subscription for UI lists. */
  readonly version = new Signal(0);
  readonly active: Signal<string>;
  private roots: LayerNode[] = [];
  private index = new Map<string, LayerNode>();
  private parents = new Map<string, LayerNode | null>();

  constructor(init: LayerInit[], activeId: string) {
    this.roots = init.map((n) => this.build(n, null));
    this.active = new Signal(activeId);
  }

  private build(n: LayerInit, parent: LayerNode | null): LayerNode {
    // Ids read from a file ("layer-12") keep the counter ahead of them, so new layers never collide.
    const m = n.id ? /^layer-(\d+)$/.exec(n.id) : null;
    if (m) uid = Math.max(uid, Number(m[1]));
    const node: LayerNode = {
      id: n.id ?? `layer-${++uid}`,
      name: n.name,
      type: n.type ?? (n.children ? 'group' : 'layer'),
      visible: n.visible ?? true,
      locked: n.locked ?? false,
      expanded: n.expanded ?? true,
      style: { ...defaultStyle, ...n.style },
      children: [],
    };
    this.index.set(node.id, node);
    this.parents.set(node.id, parent);
    node.children = (n.children ?? []).map((c) => this.build(c, node));
    return node;
  }

  get tree(): readonly LayerNode[] {
    return this.roots;
  }

  /**
   * Replaces the whole tree (a drawing opened from a file). The active
   * layer is kept when it exists as a layer, else the first layer.
   */
  reset(init: readonly LayerInit[], activeId: string): void {
    this.index.clear();
    this.parents.clear();
    this.roots = init.map((n) => this.build(n, null));
    const active = this.get(activeId)?.type === 'layer' ? activeId : (this.leaves()[0]?.id ?? activeId);
    this.active.set(active);
    this.events.emit('structure', undefined);
    this.events.emit('state', { ids: this.leaves().map((l) => l.id) });
    this.version.update((v) => v + 1);
  }

  get(id: string): LayerNode | undefined {
    return this.index.get(id);
  }

  parentOf(id: string): LayerNode | null {
    return this.parents.get(id) ?? null;
  }

  /** Leaf layers in draw order (tree order, top of list drawn last = on top). */
  leaves(): LayerNode[] {
    const out: LayerNode[] = [];
    const walk = (nodes: LayerNode[]) => {
      for (const n of nodes) n.type === 'layer' ? out.push(n) : walk(n.children);
    };
    walk(this.roots);
    return out;
  }

  leavesOf(id: string): LayerNode[] {
    const n = this.get(id);
    if (!n) return [];
    if (n.type === 'layer') return [n];
    const out: LayerNode[] = [];
    const walk = (x: LayerNode) => (x.type === 'layer' ? out.push(x) : x.children.forEach(walk));
    walk(n);
    return out;
  }

  isVisible(id: string): boolean {
    for (let n: LayerNode | null | undefined = this.get(id); n; n = this.parentOf(n.id)) if (!n.visible) return false;
    return true;
  }

  isLocked(id: string): boolean {
    for (let n: LayerNode | null | undefined = this.get(id); n; n = this.parentOf(n.id)) if (n.locked) return true;
    return false;
  }

  path(id: string): string {
    const names: string[] = [];
    for (let n: LayerNode | null | undefined = this.get(id); n; n = this.parentOf(n.id)) names.unshift(n.name);
    return names.join(' / ');
  }

  setVisible(id: string, visible: boolean): void {
    const n = this.get(id);
    if (!n || n.visible === visible) return;
    n.visible = visible;
    this.changedState(id);
  }

  toggleVisible(id: string): void {
    const n = this.get(id);
    if (n) this.setVisible(id, !n.visible);
  }

  toggleLocked(id: string): void {
    const n = this.get(id);
    if (!n) return;
    n.locked = !n.locked;
    this.changedState(id);
  }

  /** Show only this node (and its ancestors); hide every other leaf. */
  isolate(id: string): void {
    // An unknown id hides nothing (docs/adr/0020).
    if (!this.get(id)) return;
    const keep = new Set<string>();
    for (let n: LayerNode | null | undefined = this.get(id); n; n = this.parentOf(n.id)) keep.add(n.id);
    for (const l of this.leavesOf(id)) keep.add(l.id);
    for (const n of this.index.values()) n.visible = keep.has(n.id) || this.isDescendant(n.id, id);
    this.events.emit('state', { ids: this.leaves().map((l) => l.id) });
    this.version.update((v) => v + 1);
  }

  showAll(): void {
    // Everything already shown: nothing changes, not an edit (docs/adr/0020).
    if ([...this.index.values()].every((n) => n.visible)) return;
    for (const n of this.index.values()) n.visible = true;
    this.events.emit('state', { ids: this.leaves().map((l) => l.id) });
    this.version.update((v) => v + 1);
  }

  setExpanded(id: string, expanded: boolean): void {
    const n = this.get(id);
    if (!n || n.expanded === expanded) return;
    n.expanded = expanded;
    this.events.emit('expanded', { id });
    this.version.update((v) => v + 1);
  }

  setStyle(id: string, style: Partial<LayerStyle>): void {
    const n = this.get(id);
    if (!n) return;
    Object.assign(n.style, style);
    this.changedState(id);
  }

  /** Sets a layer's whole style (undo and redo restore it exactly). */
  replaceStyle(id: string, style: LayerStyle): void {
    const n = this.get(id);
    if (!n) return;
    n.style = structuredClone(style);
    this.changedState(id);
  }

  rename(id: string, name: string): void {
    const n = this.get(id);
    // The same name again changes nothing, not an edit (docs/adr/0020).
    if (!n || !name.trim() || name.trim() === n.name) return;
    n.name = name.trim();
    this.events.emit('structure', undefined);
    this.version.update((v) => v + 1);
  }

  setActive(id: string): void {
    const n = this.get(id);
    if (n?.type === 'layer') this.active.set(id);
  }

  add(init: LayerInit, parentId?: string | null): LayerNode {
    const parent = parentId ? this.get(parentId) : null;
    const container = parent && parent.type === 'group' ? parent : parent ? this.parentOf(parent.id) : null;
    const node = this.build(init, container);
    (container ? container.children : this.roots).push(node);
    if (container) container.expanded = true;
    this.events.emit('structure', undefined);
    this.version.update((v) => v + 1);
    return node;
  }

  uniqueName(base: string): string {
    const names = new Set([...this.index.values()].map((n) => n.name));
    if (!names.has(base)) return base;
    for (let i = 2; ; i++) if (!names.has(`${base} ${i}`)) return `${base} ${i}`;
  }

  private isDescendant(id: string, ancestorId: string): boolean {
    for (let n = this.parentOf(id); n; n = this.parentOf(n.id)) if (n.id === ancestorId) return true;
    return false;
  }

  private changedState(id: string): void {
    this.events.emit('state', { ids: this.leavesOf(id).map((l) => l.id) });
    this.version.update((v) => v + 1);
  }
}

File: crates/native/domain/src/layers.rs
//! The layer tree (`LayerStore` in apps/web/src/model/layers.ts): groups and
//! layers, their visibility, lock and fold, and the active layer.
//!
//! Visibility and lock are inherited: a layer is shown only when it and every
//! group above it are shown, and locked when it or a group above it is. The
//! tree is kept exactly as the file has it (`LayerNode`), so a drawing writes
//! back what it read.

use std::collections::{HashMap, HashSet};

use kentos_contracts::{LayerNode, LayerNodeType, LayerStyle, LineType};

#[derive(Clone, Debug)]
pub struct LayerTree {
    roots: Vec<LayerNode>,
    active: String,
    /// Where each node is: child indices from the roots. An id a file repeats
    /// resolves to its last node in tree order, as the web's index does.
    paths: HashMap<String, Vec<usize>>,
    /// The number in the last `layer-N` id given or read: new layers count on
    /// from it, so an id is never given twice (the web's `uid`).
    counter: u64,
}

/// A layer or group to add (the web's `LayerInit` for `LayerStore.add`).
#[derive(Clone, Debug, PartialEq)]
pub struct NewLayer {
    /// Kept when the tree has no node with it (an import's `import-parsel`);
    /// otherwise the node gets the next `layer-N`.
    pub id: Option<String>,
    pub name: String,
    pub kind: LayerNodeType,
    pub visible: bool,
    pub locked: bool,
    pub style: LayerStyle,
}

impl NewLayer {
    /// A shown, unlocked layer in the default style.
    pub fn layer(name: impl Into<String>) -> Self {
        Self {
            id: None,
            name: name.into(),
            kind: LayerNodeType::Layer,
            visible: true,
            locked: false,
            style: default_style(),
        }
    }

    /// An empty, shown group.
    pub fn group(name: impl Into<String>) -> Self {
        Self {
            kind: LayerNodeType::Group,
            ..Self::layer(name)
        }
    }
}

/// A new node's style when none is given (the web's `defaultStyle`).
pub fn default_style() -> LayerStyle {
    LayerStyle {
        color: "fg".into(),
        line_type: LineType::Continuous,
        line_weight: 0.18,
        fill: None,
        point: None,
        label: None,
        pick_interior: None,
        renderer: None,
    }
}

/// `N` of a `layer-N` id.
fn counted(id: &str) -> Option<u64> {
    id.strip_prefix("layer-")
        .filter(|n| !n.is_empty() && n.bytes().all(|b| b.is_ascii_digit()))
        .and_then(|n| n.parse().ok())
}

impl LayerTree {
    /// A tree read from a file. The active layer is kept when it is a layer,
    /// else the first layer is active (web: `LayerStore.reset`).
    pub(crate) fn new(roots: Vec<LayerNode>, active: &str) -> Self {
        let mut paths = HashMap::new();
        index(&roots, &mut Vec::new(), &mut paths);
        // Ids read from a file (`layer-12`) keep the counter ahead of them (web `build`).
        let counter = paths.keys().filter_map(|id| counted(id)).max().unwrap_or(0);
        let mut tree = Self {
            roots,
            active: String::new(),
            paths,
            counter,
        };
        tree.active = match tree.get(active) {
            Some(node) if node.kind == LayerNodeType::Layer => active.to_owned(),
            _ => tree
                .leaves()
                .first()
                .map_or_else(|| active.to_owned(), |node| node.id.clone()),
        };
        tree
    }

    /// The top of the tree, as a file writes it.
    pub fn nodes(&self) -> &[LayerNode] {
        &self.roots
    }

    /// The layer new objects go to.
    pub fn active(&self) -> &str {
        &self.active
    }

    pub fn get(&self, id: &str) -> Option<&LayerNode> {
        self.paths.get(id).and_then(|path| node(&self.roots, path))
    }

    /// The group a node is in; `None` at the top of the tree or for an unknown id.
    pub fn parent(&self, id: &str) -> Option<&LayerNode> {
        let path = self.paths.get(id)?;
        node(&self.roots, path.get(..path.len().checked_sub(1)?)?)
    }

    /// Layers (not groups) in tree order.
    pub fn leaves(&self) -> Vec<&LayerNode> {
        let mut out = Vec::new();
        leaves(&self.roots, &mut out);
        out
    }

    /// Whether the node and every group above it are shown. An unknown id is shown (web).
    pub fn is_visible(&self, id: &str) -> bool {
        self.along(id).all(|node| node.visible)
    }

    /// Whether the node or a group above it is locked. An unknown id is not (web).
    pub fn is_locked(&self, id: &str) -> bool {
        self.along(id).any(|node| node.locked)
    }

    /// The node and the groups above it.
    fn along(&self, id: &str) -> impl Iterator<Item = &LayerNode> {
        let path = self.paths.get(id).map_or(&[][..], Vec::as_slice);
        (1..=path.len()).filter_map(|n| node(&self.roots, path.get(..n)?))
    }

    fn node_mut(&mut self, id: &str) -> Option<&mut LayerNode> {
        let path = self.paths.get(id)?;
        node_mut(&mut self.roots, path)
    }

    // ── Changes; whether anything happened (the document decides what is an edit) ──

    pub(crate) fn set_visible(&mut self, id: &str, visible: bool) -> bool {
        match self.node_mut(id) {
            Some(node) if node.visible != visible => {
                node.visible = visible;
                true
            }
            _ => false,
        }
    }

    pub(crate) fn toggle_locked(&mut self, id: &str) -> bool {
        self.node_mut(id)
            .map(|node| node.locked = !node.locked)
            .is_some()
    }

    /// Shows the node, the groups above it and everything below it; hides the
    /// rest. An unknown id changes nothing (the web hides every node then).
    pub(crate) fn isolate(&mut self, id: &str) -> bool {
        let Some(target) = self.paths.get(id).cloned() else {
            return false;
        };
        visit_mut(&mut self.roots, &mut Vec::new(), &mut |path, node| {
            // The node's own path and its groups' are prefixes of the target's; the nodes below it extend it.
            node.visible = target.starts_with(path) || path.starts_with(&target);
        });
        true
    }

    /// Shows every node; false when all were shown already (nothing changed, docs/adr/0020).
    pub(crate) fn show_all(&mut self) -> bool {
        let mut changed = false;
        visit_mut(&mut self.roots, &mut Vec::new(), &mut |_, node| {
            changed |= !node.visible;
            node.visible = true;
        });
        changed
    }

    pub(crate) fn set_expanded(&mut self, id: &str, expanded: bool) -> bool {
        match self.node_mut(id) {
            Some(node) if node.expanded != expanded => {
                node.expanded = expanded;
                true
            }
            _ => false,
        }
    }

    /// Renames a node; the name is trimmed and an empty one refused. Giving
    /// the same name again still counts (web).
    pub(crate) fn rename(&mut self, id: &str, name: &str) -> bool {
        let name = name.trim();
        match self.node_mut(id) {
            // The same name again changes nothing (docs/adr/0020).
            Some(node) if !name.is_empty() && node.name != name => {
                name.clone_into(&mut node.name);
                true
            }
            _ => false,
        }
    }

    /// Makes a layer the active one; groups and unknown ids are refused.
    pub(crate) fn set_active(&mut self, id: &str) -> bool {
        if self
            .get(id)
            .is_some_and(|node| node.kind == LayerNodeType::Layer)
        {
            id.clone_into(&mut self.active);
            return true;
        }
        false
    }

    pub(crate) fn replace_style(&mut self, id: &str, style: &LayerStyle) {
        if let Some(node) = self.node_mut(id) {
            node.style.clone_from(style);
        }
    }

    /// Adds a node last in `parent` when that is a group, last in the group
    /// of `parent` when that is a layer, last at the top of the tree otherwise
    /// (an unknown `parent` too). The group it goes into is opened (web
    /// `LayerStore.add`). Returns the new node's id.
    pub(crate) fn add(&mut self, new: NewLayer, parent: Option<&str>) -> String {
        let container = parent
            .and_then(|id| Some((self.get(id)?.kind, self.paths.get(id)?.clone())))
            .and_then(|(kind, path)| match kind {
                LayerNodeType::Group => Some(path),
                // A layer's group; none at the top of the tree.
                LayerNodeType::Layer => {
                    Some(path[..path.len() - 1].to_vec()).filter(|p| !p.is_empty())
                }
            });
        let id = match new.id {
            Some(id) if !self.paths.contains_key(&id) => {
                self.counter = self.counter.max(counted(&id).unwrap_or(0));
                id
            }
            _ => self.next_id(),
        };
        let node = LayerNode {
            id: id.clone(),
            name: new.name,
            kind: new.kind,
            visible: new.visible,
            locked: new.locked,
            expanded: true,
            style: new.style,
            children: Vec::new(),
        };
        match container
            .as_deref()
            .and_then(|path| node_mut(&mut self.roots, path))
        {
            Some(group) => {
                group.children.push(node);
                group.expanded = true;
            }
            None => self.roots.push(node),
        }
        self.paths.clear();
        index(&self.roots, &mut Vec::new(), &mut self.paths);
        id
    }

    /// The next `layer-N` no node has.
    fn next_id(&mut self) -> String {
        loop {
            self.counter += 1;
            let id = format!("layer-{}", self.counter);
            if !self.paths.contains_key(&id) {
                return id;
            }
        }
    }

    /// `base` when no node (layer or group) has that name, else `base 2`,
    /// `base 3`… (web `LayerStore.uniqueName`).
    pub fn unique_name(&self, base: &str) -> String {
        let mut names = HashSet::new();
        names_of(&self.roots, &mut names);
        if !names.contains(base) {
            return base.to_owned();
        }
        (2u64..)
            .map(|i| format!("{base} {i}"))
            .find(|name| !names.contains(name.as_str()))
            .unwrap_or_else(|| base.to_owned())
    }
}

fn names_of<'a>(nodes: &'a [LayerNode], out: &mut HashSet<&'a str>) {
    for node in nodes {
        out.insert(node.name.as_str());
        names_of(&node.children, out);
    }
}

fn index(nodes: &[LayerNode], prefix: &mut Vec<usize>, paths: &mut HashMap<String, Vec<usize>>) {
    for (i, node) in nodes.iter().enumerate() {
        prefix.push(i);
        paths.insert(node.id.clone(), prefix.clone());
        index(&node.children, prefix, paths);
        prefix.pop();
    }
}

fn node<'a>(roots: &'a [LayerNode], path: &[usize]) -> Option<&'a LayerNode> {
    let (first, rest) = path.split_first()?;
    rest.iter()
        .try_fold(roots.get(*first)?, |node, i| node.children.get(*i))
}

fn node_mut<'a>(roots: &'a mut [LayerNode], path: &[usize]) -> Option<&'a mut LayerNode> {
    let (first, rest) = path.split_first()?;
    rest.iter()
        .try_fold(roots.get_mut(*first)?, |node, i| node.children.get_mut(*i))
}

fn leaves<'a>(nodes: &'a [LayerNode], out: &mut Vec<&'a LayerNode>) {
    for node in nodes {
        match node.kind {
            LayerNodeType::Layer => out.push(node),
            LayerNodeType::Group => leaves(&node.children, out),
        }
    }
}

/// Calls `f` with every node and its path, in tree order.
fn visit_mut(
    nodes: &mut [LayerNode],
    prefix: &mut Vec<usize>,
    f: &mut impl FnMut(&[usize], &mut LayerNode),
) {
    for (i, node) in nodes.iter_mut().enumerate() {
        prefix.push(i);
        f(prefix, node);
        visit_mut(&mut node.children, prefix, f);
        prefix.pop();
    }
}
